File: parnamondal/Week - 2/Day - 1/Reverse Pairs.py
# Time Complexity - O(NlogN)
# Space Complexity - O(N)


class Solution:
    def reversePairs(self, nums: List[int]) -> int:
        n=len(nums)
        if n<=1:
            return 0
        def mergesort (nums,low,high,mid):
            cnt=0
            j=mid+1
            for i in range (low,mid+1):
                while(j<=high and nums[i]>nums[j]*2):
                    j+=1
                cnt+=(j-(mid+1))
            tmp=[]
            left,right=low,mid+1
            
            while (left<=mid and right<=high):
                if nums[left]<=nums[right]:
                    tmp.append(nums[left])
                    left+=1
                else:
                    tmp.append(nums[right])
                    right+=1
            while left<=mid:
                tmp.append(nums[left])
                left+=1
            while right<=high:
                tmp.append(nums[right])
                right+=1
            for i in range (low,high+1):
                nums[i]=tmp[i-low]
            return cnt
                    
                    
                    
            
                    
                    
        def merge(nums,low,high):
            if low>=high:
                return 0
            #inv=0
            mid=(low+high)//2
            inv=merge(nums,low,mid)
            inv+=merge(nums,mid+1,high)
            inv+=mergesort(nums,low,high,mid)
            return inv
            
        
        return merge(nums,0,len(nums)-1)
        l

File: Adhiti_Shaili/Day 7,8/Reverse pairs.cpp
class Solution {
public:
        
    void merge(vector<int>& nums,int s,int mid,int e)
    {
        int n1=mid-s+1;
        int n2=e-mid;
        vector<int> v1(n1,0);
        vector<int> v2(n2,0);
        
        for(int i=0;i<n1;i++)
        {
            v1[i]=nums[s+i];
        }
        
        for(int i=0;i<n2;i++)
        {
            v2[i]=nums[mid+1+i];
        }
        
        int k=s,i=0,j=0;
        while(i<n1&&j<n2)
        {
            if(v1[i]<v2[j])
            {
                nums[k++]=v1[i++];
            }
            else
            {
                nums[k++]=v2[j++];
            }
        }
        while(i<n1)
        {
            nums[k++]=v1[i++];
        }
        while(j<n2)
        {
            nums[k++]=v2[j++];
        }
    }
    
    int mergesort(vector<int>& nums,int s,int e)
    {
        if(s>=e)
        {
            return 0;
        }
        int mid=(s+e)/2;
        int count=mergesort(nums,s,mid)+mergesort(nums,mid+1,e);
        int j=mid+1;
        for(int i=s;i<=mid;i++)
        {
            while(j<=e&&(nums[i]>(long long)nums[j]*2))
            {
                j++;
            }
            count+=j-(mid+1);
        }
        merge(nums,s,mid,e);
        return count;
    }
    
    
    int reversePairs(vector<int>& nums) {
        return mergesort(nums,0,nums.size()-1);
               
    }
};